File: frontend/src/lib/grocery.ts
import { writable, derived } from 'svelte/store';
import { ingredientKey, formatQuantity } from './ingredients';
import type { ParsedIngredient } from './ingredients';
import type { GroceryList } from './types';
import {
  getGroceryList,
  addRecipeToGroceryApi,
  removeRecipeFromGroceryApi,
  toggleGroceryChecked,
  removeGroceryItem,
  clearGroceryChecked,
  clearGroceryAll,
} from './api';

export interface GroceryRecipe {
  title: string;
  fork: string | null;
  servings: string | null;
  items: ParsedIngredient[];
}

export interface GroceryStore {
  recipes: Record<string, GroceryRecipe>;
  checked: string[];
  customCombines: Array<{ keys: string[]; quantity: number | null; unit: string | null; name: string }>;
}

function emptyStore(): GroceryStore {
  return { recipes: {}, checked: [], customCombines: [] };
}

function apiToStore(data: GroceryList): GroceryStore {
  return {
    recipes: data.recipes as unknown as Record<string, GroceryRecipe>,
    checked: data.checked,
    customCombines: [],
  };
}

export const groceryStore = writable<GroceryStore>(emptyStore());

export async function initGroceryStore() {
  try {
    const data = await getGroceryList();
    groceryStore.set(apiToStore(data));
  } catch {
    // If API fails, start with empty store
  }
}

export async function addRecipeToGrocery(
  slug: string,
  title: string,
  ingredients: string[],
  fork: string | null = null,
  servings: string | null = null,
) {
  try {
    const data = await addRecipeToGroceryApi(slug, title, ingredients, fork, servings);
    groceryStore.set(apiToStore(data));
  } catch (e) {
    console.error('Failed to add recipe to grocery list', e);
  }
}

export async function removeRecipeFromGrocery(slug: string) {
  try {
    const data = await removeRecipeFromGroceryApi(slug);
    groceryStore.set(apiToStore(data));
  } catch (e) {
    console.error('Failed to remove recipe from grocery list', e);
  }
}

export async function toggleChecked(key: string) {
  try {
    const data = await toggleGroceryChecked(key);
    groceryStore.set(apiToStore(data));
  } catch (e) {
    console.error('Failed to toggle checked', e);
  }
}

export async function clearChecked() {
  try {
    const data = await clearGroceryChecked();
    groceryStore.set(apiToStore(data));
  } catch (e) {
    console.error('Failed to clear checked', e);
  }
}

export async function removeItem(key: string) {
  try {
    const data = await removeGroceryItem(key);
    groceryStore.set(apiToStore(data));
  } catch (e) {
    console.error('Failed to remove item', e);
  }
}

export async function clearAll() {
  try {
    const data = await clearGroceryAll();
    groceryStore.set(apiToStore(data));
  } catch (e) {
    console.error('Failed to clear grocery list', e);
  }
}

export function addCustomCombine(keys: string[], quantity: number | null, unit: string | null, name: string) {
  groceryStore.update(store => {
    store.customCombines.push({ keys, quantity, unit, name });
    return store;
  });
}

export interface MergedItem {
  key: string;
  quantity: number | null;
  unit: string | null;
  name: string;
  displayText: string;
  sources: string[];
  checked: boolean;
}

export function getMergedItems(store: GroceryStore): MergedItem[] {
  const map = new Map<string, { quantity: number | null; unit: string | null; name: string; sources: string[] }>();

  // Collect keys that have been custom-combined
  const combinedKeys = new Set<string>();
  for (const combine of store.customCombines) {
    for (const k of combine.keys) combinedKeys.add(k);
  }

  for (const [slug, recipe] of Object.entries(store.recipes)) {
    for (const item of recipe.items) {
      const key = ingredientKey(item);
      if (combinedKeys.has(`${slug}:${key}`)) continue;

      if (map.has(key)) {
        const existing = map.get(key)!;
        if (existing.quantity !== null && item.quantity !== null && existing.unit === item.unit) {
          existing.quantity += item.quantity;
        }
        if (!existing.sources.includes(recipe.title)) {
          existing.sources.push(recipe.title);
        }
      } else {
        map.set(key, {
          quantity: item.quantity,
          unit: item.unit,
          name: item.name,
          sources: [recipe.title],
        });
      }
    }
  }

  // Add custom combines
  for (const combine of store.customCombines) {
    const key = `custom:${combine.name}`;
    map.set(key, {
      quantity: combine.quantity,
      unit: combine.unit,
      name: combine.name,
      sources: [],
    });
  }

  const items: MergedItem[] = [];
  for (const [key, val] of map) {
    const qtyStr = val.quantity !== null ? formatQuantity(val.quantity) : '';
    const unitStr = val.unit || '';
    const displayText = [qtyStr, unitStr, val.name].filter(Boolean).join(' ');

    items.push({
      key,
      quantity: val.quantity,
      unit: val.unit,
      name: val.name,
      displayText,
      sources: val.sources,
      checked: store.checked.includes(key),
    });
  }

  return items.sort((a, b) => a.name.localeCompare(b.name));
}

export const recipeCount = derived(groceryStore, $store => Object.keys($store.recipes).length);

File: backend/app/routes/grocery.py
"""Server-side grocery list API."""

import json
import logging
from pathlib import Path

from fastapi import APIRouter, HTTPException
from fastapi.responses import PlainTextResponse

from app.ingredients import parse_ingredient, ingredient_key, format_quantity
from app.models import AddToGroceryRequest, GroceryItem, GroceryList, GroceryRecipe

logger = logging.getLogger(__name__)


def create_grocery_router(recipes_dir: Path) -> APIRouter:
    router = APIRouter(prefix="/api/grocery")
    grocery_path = recipes_dir / "grocery-list.json"

    def _load() -> GroceryList:
        if not grocery_path.exists():
            return GroceryList()
        try:
            data = json.loads(grocery_path.read_text())
            return GroceryList(**data)
        except Exception:
            logger.exception("Failed to load grocery list")
            return GroceryList()

    def _save(store: GroceryList) -> None:
        grocery_path.write_text(json.dumps(store.model_dump(), indent=2))

    @router.get("")
    def get_grocery_list():
        return _load().model_dump()

    @router.post("/recipes")
    def add_recipe_to_grocery(req: AddToGroceryRequest):
        store = _load()
        items = []
        for line in req.ingredients:
            parsed = parse_ingredient(line)
            items.append(GroceryItem(**parsed))
        store.recipes[req.slug] = GroceryRecipe(
            title=req.title,
            fork=req.fork,
            servings=req.servings,
            items=items,
        )
        _save(store)
        return store.model_dump()

    @router.delete("/recipes/{slug}")
    def remove_recipe_from_grocery(slug: str):
        store = _load()
        if slug in store.recipes:
            del store.recipes[slug]
            _save(store)
        return store.model_dump()

    @router.post("/check/{item_key:path}")
    def toggle_checked(item_key: str):
        store = _load()
        if item_key in store.checked:
            store.checked.remove(item_key)
        else:
            store.checked.append(item_key)
        _save(store)
        return store.model_dump()

    @router.delete("/items/{item_key:path}")
    def remove_item(item_key: str):
        store = _load()
        # Remove the item from all recipes
        empty_slugs = []
        for slug, recipe in store.recipes.items():
            recipe.items = [
                item for item in recipe.items
                if ingredient_key(item.model_dump()) != item_key
            ]
            if not recipe.items:
                empty_slugs.append(slug)
        for slug in empty_slugs:
            del store.recipes[slug]
        store.checked = [k for k in store.checked if k != item_key]
        _save(store)
        return store.model_dump()

    @router.delete("/checked")
    def clear_checked():
        store = _load()
        store.checked = []
        _save(store)
        return store.model_dump()

    @router.delete("")
    def clear_all():
        store = GroceryList()
        _save(store)
        return store.model_dump()

    @router.get("/export")
    def export_grocery():
        store = _load()
        # Merge items across recipes
        merged: dict = {}
        for recipe in store.recipes.values():
            for item in recipe.items:
                key = ingredient_key(item.model_dump())
                if key in merged:
                    existing = merged[key]
                    if existing["quantity"] is not None and item.quantity is not None and existing["unit"] == item.unit:
                        existing["quantity"] += item.quantity
                else:
                    merged[key] = {
                        "quantity": item.quantity,
                        "unit": item.unit,
                        "name": item.name,
                        "displayText": item.displayText,
                    }

        unchecked_lines = []
        checked_lines = []
        for key, val in sorted(merged.items(), key=lambda kv: kv[1]["name"]):
            qty_str = format_quantity(val["quantity"]) if val["quantity"] is not None else ""
            unit_str = val["unit"] or ""
            display = " ".join(filter(None, [qty_str, unit_str, val["name"]]))
            if key in store.checked:
                checked_lines.append(f"[x] {display}")
            else:
                unchecked_lines.append(f"[ ] {display}")

        lines = []
        if unchecked_lines:
            lines.append("To buy:")
            lines.extend(unchecked_lines)
        if checked_lines:
            if lines:
                lines.append("")
            lines.append("Got it:")
            lines.extend(checked_lines)

        if not lines:
            lines.append("Grocery list is empty.")

        return PlainTextResponse(
            content="\n".join(lines) + "\n",
            headers={"Content-Disposition": 'attachment; filename="grocery-list.txt"'},
        )

    return router
